### PR TITLE
 Fix crash bug with duplicate inputs within a transaction

### PR DESCRIPTION
**Pull request by TheBlueMatt**

### CODE REVIEW

**Comment by laanwj** [comment]
tested ACK 9b4a36effcf642f3844c6696b757266686ece11a

**Comment by jamesob** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14247/commits/9b4a36effcf642f3844c6696b757266686ece11a

**Comment by sdaftuar** [comment]
ACK

**Comment by MarcoFalke** [comment]
utACK 9b4a36effcf642f3844c6696b757266686ece11a. Checked that the test fails without the fix and passes with the fix.

File: src/validation.cpp [hunk 1]
```diff
@@ -3122,7 +3122,7 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
 
     // Check transactions
     for (const auto& tx : block.vtx)
-        if (!CheckTransaction(*tx, state, false))
+        if (!CheckTransaction(*tx, state, true))
```
Comment (ryanofsky) [inline src/validation.cpp]: Just noting it looks like this argument can be removed entirely in a future cleanup PR.

**Review by ryanofsky — COMMENTED** [review]
Slightly tested ACK 9b4a36effcf642f3844c6696b757266686ece11a, just ran python test with & without fix.

**Comment by gmaxwell** [comment]
ACK

**Comment by practicalswift** [comment]
utACK 9b4a36effcf642f3844c6696b757266686ece11a

Very nice find @sdaftuar! How did you find this issue?

**Comment by luke-jr** [comment]
For reference, this issue was assigned CVE-2018-17144

**Comment by ftrader** [comment]
Could @TheBlueMatt / @sdaftuar outline the responsible disclosure procedures, if any, that were attempted to other projects before publishing these pull requests?

**Comment by TheBlueMatt** [comment]
The bug was disclosed to other projects simultaneously to it being disclosed to us. We tried to coordinate timelines for release but sadly didn't want to wait much past EST business hours to get people patched so ran out of time.

**Comment by zquestz** [comment]
Where can I find a detailed description of the impact of this bug? When is the CVE getting published so I can get more clarity on the exact impact? I am seeing all sorts of reports on Twitter and would like to get an accurate assessment of the issue.

**Comment by PierreRochard** [comment]
> Where can I find a detailed description of the impact of this bug? When is the CVE getting published so I can get more clarity on the exact impact? I am seeing all sorts of reports on Twitter and would like to get an accurate assessment of the issue.

Release notes: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.16.3.md#denial-of-service-vulnerability

More color in today's OpTech newsletter ( https://mailchi.mp/cc96f82565eb/bitcoin-optech-newsletter-13?e=cbe2b70569 ):

Upgrade to Bitcoin Core 0.16.3 to fix denial-of-service vulnerability: a bug introduced in Bitcoin Core 0.14.0 and affecting all subsequent versions through to 0.16.2 will cause Bitcoin Core to crash when attempting to validate a block containing a transaction that attempts to spend the same input twice. Such blocks would be invalid and so can only be created by miners willing to lose the allowed income from having created a block (at least 12.5 XBT or $80,000 USD).

Patches for master and 0.16 branches were submitted for public review yesterday, the 0.16.3 release has been tagged containing the patch, and binaries will be available for download as soon as a sufficient number of well-known contributors have reproduced the deterministic build—probably later today (Tuesday). Immediate upgrade is highly recommended.

**Comment by sickpig** [comment]
> We tried to coordinate timelines for release but sadly didn't want to wait much past EST business hours to get people patched so ran out of time.

I was among the people who received the anonymous report, even though BU was not affected, and I'm not aware of any kind of effort to "coordinate timelines".

As far as I can tell the list of people to which the bug was disclosed was not exhaustive, not by a long shot, so make sure to have a proper way to tell projects/devs that their code is at risk is a must IMHO. 

It would be great to know what were the actions taken to put in places the aforementioned coordination process. The aim is to improve the process in case of new disclosures that could happen in the future.

**Comment by Sjors** [comment]
Post merge slightly tested 9b4a36e, ran python test without fix which produced the assert.

**Comment by biblepay** [comment]
Thanks.

**Comment by practicalswift** [comment]
I would like to thank the researcher who invested the time into find this issue. Impressive finding! Thanks!

**Comment by isghe** [comment]
Is this bug reproducible in regtest or testnet? Thanks

**Comment by Sjors** [comment]
@isghe the bug is reproducible on regtest using the Python test from this PR. You need to change `CheckTransaction(*tx, state, true)` back to `CheckTransaction(*tx, state, false)` and then run `test/functional/p2p_invalid_block.py`. The test contains the kind of invalid block that can crash a node. You'll see a crash caused by an `assert`.

CVE-2018-17144 Full Disclosure, which includes the inflation bug not mentioned above: https://github.com/bitcoin-core/bitcoincore.org

I assume we'll add a test case for the inflation bug once the dust has settled?

**Comment by promag** [comment]
> I assume we'll add a test case for the inflation bug once the dust has settled?

@gmaxwell already suggested it, also agree.

**Comment by ftrader** [comment]
The discoverer of CVE-2018-17144 has come forward (for those wishing to express their gratitude, there is an included donation address in the post):

https://medium.com/@awemany/600-microseconds-b70f87b0b2a6

**Comment by isghe** [comment]
It looks they created with success 0.1 tBTC from nothing, using this bug in testnet3 block 1414411.

**Pull request merged**